add C day 3 and M4 days 1-7

ebblake opened the pull request:
The trailing newlines are essential parts of the files provided by AoC.  Without them, at least the C day 3 submission dumps core.

ebblake committed:
guarantee trailing newline for C

The trailing newlines are essential parts of the files provided by AoC
(even for things like 2019 day 4 where there is no explicit download
link, the URL https://adventofcode.com/2019/day/4/input still works
and includes the newline).  Without them, at least the upcoming C day
3 submission dumps core.

ebblake committed:
golf a couple bytes off C days 1,2

No need to print a trailing newline

Starwort commented:
You can make the runner for C add a newline (`echo | cat $day.in - | prog`), but I disagree that the trailing newline is an essential part of the file; it isn't part of the *data* (especially for days where your input is provided in-line, such as [2019 day 4](https://adventofcode.com/2019/day/4)). It's totally fine to make it part of the language-specific rules that inputs are concatenated with a newline, though. Would merge if this is moved to a language-specific rule rather than a global one (the change would break several existing Python entries which depend on the input being exactly the correct size)

ebblake committed:
M4 solutions days 1-7 as posted on megathreads

As with C, the trailing newline is essential to some of my golfing.
However, unlike C, m4 can't take its input on stdin.  While I test
with -Di=file locally, for this repo it's easier to create an actual
file 'i' with the right contents.

ebblake commented:
For 2019 day 4, you can STILL download the file at https://adventofcode.com/2019/day/4/input, and that download DOES have the trailing newline.

At any rate, it is essential to several of my golfed solutions (at least in C and m4), so I'll be adding it in the runners for those languages and resubmitting.

ebblake committed:
C submission for day 3

The state machine here requires the trailing newline.
